### PR TITLE
Added Boxcar 2

### PR DESCRIPTION
**Pull request by srod**
Hi,

I added boxcar.io support, related to #322 

I have duplicated PushBullet because it was matching the boxcar API.
With somes changes to fit the boxcar API.

I also added a new response status code to RestSharpExtension.
Because boxcar API send back a 201 Created.

Thank you for your code review ;)

### CODE REVIEW

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            var request = new RestRequest(Method.POST);
+
+            try
+            {
+                SendNotification(title, message, request, settings);
+            }
+            catch (BoxcarException ex)
+            {
+                _logger.ErrorException("Unable to send message", ex);
+                error = true;
```
Comment (kayone) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: why not just throw the exception here?

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        public ValidationFailure Test(BoxcarSettings settings)
+        {
+            try
+            {
+                const string title = "Test Notification";
+                const string body = "This is a test message from Sonarr";
+
+                SendNotification(title, body, settings);
```
Comment (kayone) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: return null here.

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            }
+            catch (RestException ex)
+            {
+                if (ex.Response.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    _logger.ErrorException("Access Token is invalid: " + ex.Message, ex);
+                    throw;
+                }
+
+                throw new BoxcarException("Unable to send text message: {0}", ex, ex.Message);
```
Comment (kayone) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: I think you mean `ex.Message, ex`

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs (hunk 1 above)
Comment (srod) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: Done

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs (hunk 2 above)
Comment (srod) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: Done

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs (hunk 3 above)
Comment (srod) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: The build was failing with ```throw new BoxcarException("Unable to send text message: {0}", ex.Message, ex);```
So I end up with ```throw new BoxcarException("Unable to send text message: " + ex.Message, ex);```

**Comment by srod** [comment]
Is everything ok, or I missed something ?

**Comment by kayone** [comment]
sorry about the delay, I don't have an ios device to test it, @markus101 @Taloth can you guys take a look if you have one.

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarSettings.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            RuleFor(c => c.Token).NotEmpty();
+        }
+    }
+
+    public class BoxcarSettings : IProviderConfig
+    {
+        private static readonly BoxcarSettingsValidator Validator = new BoxcarSettingsValidator();
+
+        [FieldDefinition(0, Label = "Access Token", HelpLink = "https://boxcar.io/client")]
```
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarSettings.cs]: Needs HelpText with information on where to get the key from (Settings in the iOS app), the HelpLink isn't useful here.

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    public interface IBoxcarProxy
+    {
+        void SendNotification(string title, string message, BoxcarSettings settings);
+        ValidationFailure Test(BoxcarSettings settings);
+    }
+
+    public class BoxcarProxy : IBoxcarProxy
+    {
+        private readonly Logger _logger;
+        private const string URL = "https://new.boxcar.io/api/notifications";
```
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: This URL gives me a 404.

**Comment by markus101** [comment]
A couple things to clean up and the code needs to be rebased off of our develop branch. There are a couple tweaks that will likely need to be made as the schema has changed slightly and you'll need to disable the on rename handler: https://github.com/Sonarr/Sonarr/blob/develop/src/NzbDrone.Core/Notifications/PushBullet/PushBullet.cs#L48-L54

Let us know if you need help with this and we can take a look.

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarSettings.cs (hunk 4 above)
Comment (srod) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarSettings.cs]: I added a HelpText and change the link to the right page, where we can find the token.

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs (hunk 5 above)
Comment (srod) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: The URL is in 404 with a GET request.
You have to do a POST on this url with parameters :
https://new.boxcar.io/api/notifications?user_credentials=accessToken&notification[title]=testTitle&notification[long_message]=testMessage

**Comment by srod** [comment]
I think all is OK.
I added a HelpText to the token, to be more explicit.
And I have rebased the branch and changed the rename Handler.

File: src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs (hunk 5 above)
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Boxcar/BoxcarProxy.cs]: Sorry, you're right, I was thinking this was the link from the notification to the site so users could see what it was. Makes sense.

**Comment by markus101** [comment]
Awesome, I think that covers everything. Could you squash everything into one commit: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

`git rebase -i HEAD~5` should do it. You'll pick the first commit and squash the other 4 into it. Make the commit message `New: Boxcar 2 notifications`.

**Comment by srod** [comment]
Ok, it's done :)

**Comment by markus101** [comment]
Thanks for adding this! Merged :smile:

**Comment by srod** [comment]
Cool ! :D

**Pull request merged**